### ISSUE TITLE
MP4 "free" chunks can cause very slow parsing

### ISSUE THREAD

**Issue by LeoDavidson** [MediaArea/MediaInfoLib#1353]
I ran into a couple of small (~80 MB) MP4 files which took about 30 seconds each (in debug builds) for MediaInfoLib to parse, compared to 1-2 seconds for normal files.

I can send them privately if you want. Just let me know where to upload them.

After some investigation, it appears to be because they have "free" chunks/blocks which MediaInfoLib parses in a very slow way. It seems to skip 4 bytes of them at a time, and the blocks are quite large in these files. The files also seem to have a lot of the "free" blocks for some reason, which results in repetitive calls to ZtringListList::Find. Since that Find method does linear lookups, with about 300 strings in the list, it accounted for a lot of the processing time.

I found what seems to be a fix, but am not sure enough that it is correct, so I wanted to raise this as an issue rather than a pull request. (I can make this a pull request if you want, though, of course!)

My fix was to change moov_udta_xxxx() so it returns immediately if it's a "free" block:

```
void File_Mpeg4::moov_udta_xxxx()
{
    if (Element_Code == Elements::free)
        return;

    //Getting the method
    ...
```

That makes the files process very quickly and I haven't noticed any side effects, but I don't know the code well enough to be sure it's the right solution.

Before doing that, I also tried making ZtringListList::Find check if it was being asked to find "free" and fail instantly, as a quick test, and that improved speed quite a bit, but not as much as skipping the "free" blocks entirely. So the repeated linear lookups were slowing things down a lot, but other things were involved as well.

(I was thinking about making ZtringListList cache the previous Find to speed up cases where it was called for the same thing again, or adding an std::map for faster lookups, but it doesn't seem needed with the moov_udta_xxxx change skipping "free" blocks entirely. Maybe there are other situations where those would help, but maybe the overheads of maintaining the extra data would cost more in typical usage. I don't know the project well enough to say.)

Many thanks for your time!

**Comment by LeoDavidson** [comment]
I just realised I hadn't tried these file with a standard/official release of MediaInfo.

Just downloaded the Windows build of MediaInfo GUI 20.09 and it opens both files instantly.

The problem might only happen with certain compilation flags defined, in that case. If you're interested, I can try to track down which are doing it.

**Comment by JeromeMartinez** [comment]
I didn't expect to get `free` atoms in metadata, but it makes sense (edition etc). Even if not visible in release mode, it makes sense to sip such atom.
You may add the `free` atom in the list in `File_Mpeg4::Data_Parse()`, and map it to `File_Mpeg4::free()`.
It is quick to do for who knows the code, don't hesitate to drop me an email with a link to download privately the file or I may send you a private FTP server for upload.

**Comment by LeoDavidson** [comment]
Many thanks! I've uploaded the files to a web server and emailed the link to you. Give me a shout if the mail doesn't arrive.

**Comment by JeromeMartinez** [comment]
I was trying to parse the big `free` atom as if this is a sequence of tag elements, and as it is a full of zeroes the parser was slow... In debug mode only, because the debug mode is slow with the underlying function calls.
But... This code is actually skipped in the current dev version.
I did a quick and dirty fix https://github.com/MediaArea/MediaInfoLib/pull/1354 for being sure that even if 1 day I want to parse such atoms I don't do the same mistake.
Note that metadata present in the file (the "Xtra" atom from Microsoft) is not parsed, I don't have the time for the moment for implementing that, maybe one day...

**Comment by LeoDavidson** [comment]
I can confirm this works on my test files and my local build/setup, after removing my hacky fix and merging just the #1354 changes into 20.09. Looks good!
